homeassistant/components/modbus/binary_sensor.py

```diff
@@ -61,10 +78,41 @@ async def async_update(self, now: datetime | None = None) -> None:
                 return
             self._lazy_errors = self._lazy_error_count
             self._attr_available = False
-            self.async_write_ha_state()
-            return
+            result_bits = None
+        else:
+            self._lazy_errors = self._lazy_error_count
+            self._attr_is_on = result.bits[0] & 1
+            self._attr_available = True
+            result_bits = result.bits
 
-        self._lazy_errors = self._lazy_error_count
-        self._attr_is_on = result.bits[0] & 1
-        self._attr_available = True
+        self.schedule_update_ha_state()
         self.async_write_ha_state()
+        for entry in self._slaves:
+            entry.set_from_master(result_bits, self._attr_available)
+
+
+class SlaveSensor(RestoreEntity, BinarySensorEntity):
+    """Modbus slave binary sensor."""
+
+    def __init__(self, inx: int, entry: dict[str, Any]) -> None:
+        """Initialize the Modbus binary sensor."""
+        self._attr_name = f"{entry[CONF_NAME]}_{inx}"
+        self._attr_device_class = entry.get(CONF_DEVICE_CLASS)
+        self._attr_should_poll = False
+        self._attr_available = False
+        self._result_inx = int(inx / 8)
+        self._result_bit = 2 ** (inx % 8)
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        if state := await self.async_get_last_state():
+            self._attr_is_on = state.state == STATE_ON
+            self._attr_available = True
+            self.async_write_ha_state()
+
+    def set_from_master(self, bits: list[int] | None, available: bool) -> None:
+        """Update the state of the sensor."""
+        if bits is not None:
+            self._attr_is_on = bits[self._result_inx] & self._result_bit
+        self._attr_available = available
+        self.schedule_update_ha_state()
```

Comment:
Called from async context. 
```suggestion
        self.async_write_ha_state()
```

Reply:
updated.